### PR TITLE
nabupc/keyboard: Fully implement true keyboard emulation on the NABU

### PR DESCRIPTION
**Pull request by brijohn**
I finally had time to do a bit more in depth analysis of one of my NABU keyboards and was able to figure out the missing information on how everything fits together.

This patch is the result of that and gets the existing stub for low level keyboard emulation up and working. This includes adding the keyboard matrix as well as support for the joysticks and paddles.

### CODE REVIEW

File: src/devices/bus/nabupc/keyboard/keyboard.cpp [hunk 1]
```diff
@@ -80,40 +262,54 @@ keyboard_device::keyboard_device(machine_config const &mconfig, char const *tag,
...
+	m_adc->in_callback<7>().set_ioport("PADDLE8");
 }
 
 void keyboard_device::nabu_kb_mem(address_map &map)
 {
 	map(0x5000, 0x5300).r(FUNC(keyboard_device::gameport_r));
+	map(0x7000, 0x7000).w(FUNC(keyboard_device::adc_start_w));
+	map(0x9000, 0x9000).w(FUNC(keyboard_device::cpu_ack_irq_w));
+	map(0xB000, 0xB700).w(FUNC(keyboard_device::adc_latch_w));
+	map(0xD000, 0xD000).r(m_adc, FUNC(adc0809_device::data_r));
```
Comment (cuavas) [inline src/devices/bus/nabupc/keyboard/keyboard.cpp:312]: Please use lowercase hexadecimal digits.

**Review by cuavas — COMMENTED** [review]
Since there’s only one keyboard, can the bus be removed and the device be hooked up with the signal handlers directly like other systems that only have a single keyboard option (e.g. Kaypro or Zorba)?  We shouldn’t need the simulated keyboard either now that the actual keyboard is emulated.

File: src/devices/bus/nabupc/keyboard/keyboard.cpp [hunk 2]
```diff
@@ -68,6 +104,152 @@ ROM_START(nabu_keyboard_rom)
...
+	PORT_BIT( 0x80, IP_ACTIVE_LOW, IPT_BUTTON1 ) PORT_PLAYER(3)               PORT_CODE(INPUT_CODE_INVALID)
+	PORT_START("JOYSTICK4")
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_JOYSTICK_LEFT ) PORT_PLAYER(4)         PORT_CODE(INPUT_CODE_INVALID)
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_JOYSTICK_DOWN ) PORT_PLAYER(4)         PORT_CODE(INPUT_CODE_INVALID)
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_JOYSTICK_RIGHT ) PORT_PLAYER(4)        PORT_CODE(INPUT_CODE_INVALID)
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_JOYSTICK_UP ) PORT_PLAYER(4)           PORT_CODE(INPUT_CODE_INVALID)
+	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x40, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x80, IP_ACTIVE_LOW, IPT_BUTTON1 ) PORT_PLAYER(4)               PORT_CODE(INPUT_CODE_INVALID)
```
[the remark this answers is not in the document]
Comment (brijohn) [reply src/devices/bus/nabupc/keyboard/keyboard.cpp:232]: But if i don't set some default values here then when it starts up it gets assigned various keyboard values that conflict with the keyboard mappings, which is equally annoying and I find unexpected. It seems better to have a default unmapped state and then allow people to configure joysticks and paddles for the system as needed so you don't end up with conflicting mappings.

[the remark this answers is not in the document]
Comment (cuavas) [reply src/devices/bus/nabupc/keyboard/keyboard.cpp]: I know the default general input assignments conflict with keyboards, but anyone who actually uses computers emulation in MAME will set their general input assignments to something on a game controller once and then expect it to actually work across systems.  You also break controller configuration files when you do this.  I’ll remove the fixed assignments myself.

**Pull request merged**